axum-login/src/middleware.rs

```diff
@@ -141,6 +174,27 @@ macro_rules! permission_required {
 /// used as the response.
 #[macro_export]
 macro_rules! predicate_required {
+    ($predicate:expr, failed_predicate_callback = $failed_predicate_callback:expr) => {{
+        use $crate::axum::{
+            extract::OriginalUri,
+            middleware::{from_fn, Next},
+            response::IntoResponse,
+        };
+
+        from_fn(
+            |auth_session: $crate::AuthSession<_>,
+             OriginalUri(original_uri): OriginalUri,
+             req,
+             next: Next| async move {
+                if $predicate(auth_session).await {
+                    next.run(req).await
+                } else {
+                    $failed_predicate_callback(original_uri).await.into_response()
```

Comment:
I think this is redundant with the branch below, right? Maybe we could update $alternative to take req in that case.

One reason I didn't suggest this approach is that the next parameter is not easily accessible in this particular branch. That makes handling successful redirects more difficult.

Reply:
so you are suggesting something like this. and remove $failed_predicate_callback branch?

```

    ($predicate:expr, $alternative:expr) => {{
        use $crate::axum::{
            middleware::{from_fn, Next},
            response::IntoResponse,
        };

        from_fn(
            |auth_session: $crate::AuthSession<_>, req, next: Next| async move {
                if $predicate(auth_session).await {
                    next.run(req).await
                } else {
                    $alternative(req).await.into_response()
                }
            },
        )
    }};
    ```

Reply:
Something like that, but then handling the query params seems dicey.

Reply:
When failed_predicate_callback it's up to the user to handle the redirect URL. So I don't think there is a need to handle query parameters.